### ISSUE TITLE
Investigate Hive SyncSets

### ISSUE THREAD

**Issue by djzager** [openshift/hive#801]
# Test Setup

To make adopting the minikube cluster easier, it's best to use `minikube config set embed-certs true` so that created clusters have their certificates embedded in the kubeconfig file. The setup that follows for the management and managed cluster apply for each scenario.

## Management Cluster (hub)

1. Create cluster `minikube start -p hub`
2. Use hub context `kubectl config use-context hub`
3. Checkout `v1` branch of hive, `make clean && make build`
4. Deploy hive, `make deploy`
5. Deploy admission-certs `rm -r hiveadmission-certs && ./hack/hiveadmission-dev-cert.sh`

## Managed Cluster (cluster0)

1. Create cluster `minikube start -p cluster0`
2. Use cluster0 context `kubectl config use-context cluster0`
3. Create BareMetalHosts CRD `kubectl --context cluster0 apply -f https://raw.githubusercontent.com/metal3-io/baremetal-operator/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml`
4. Create `openshift-machine-api` namespace `kubectl --context cluster0 create namespace openshift-machine-api`
5. Use hub context `kubectl config use-context hub`
6. Adopt cluster, `bin/hiveutil create-cluster --adopt --adopt-admin-kubeconfig=cluster0.config --include-secrets=false --adopt-cluster-id=cluster0 --adopt-infra-id=cluster0 cluster0` (the `cluster0.config` was created using a copy of `$HOME/.kube/config` to only include information for cluster0.

# Test Scenarios

## SyncSet Cross-Namespaced Secret

In this scenario we want to see if a SyncSet can reference a Secret in a different namespace

### **Steps**

1. Create a 'syncset-secret-test' namespace `kubectl --context hub create namespace syncset-secret-test`
2. Create a Secret `kubectl --context hub create secret generic --namespace syncset-secret-test syncset-secret-test --from-literal=username='exampleuser' --from-literal=password='examplepassword'`
3. Create a SyncSet:

        $ cat secret-syncset-test.cr.yaml
        ---
        apiVersion: hive.openshift.io/v1
        kind: SyncSet
        metadata:
          name: syncset-secret-test
          namespace: default
        spec:
          clusterDeploymentRefs:
          - name: cluster0
          resourceApplyMode: Sync
          secretMappings:
          - sourceRef:
              name: syncset-secret-test
              namespace: syncset-secret-test
            targetRef:
              name: syncset-secret-test
              namespace: default
        
        $ kubectl --context hub create -f secret-syncset-test.cr.yaml

### Expected Result

The `syncset-secret-test` secret to be created in the `default` namespace of the managed (`cluster0`) cluster

### Actual Result

    $ kubectl --context cluster0 -n default get secret syncset-secret-test -o go-template="{{.data.username}}" | base64 --decode
    exampleuser%
    
    $ kubectl --context cluster0 -n default get secret syncset-secret-test -o go-template="{{.data.password}}" | base64 --decode
    examplepassword%

## SyncSet Merges Labels

In this scenario, a BareMetalHost resource already exists on the managed cluster with some labels. We will create a SyncSet with a BareMetalHost resource that 1) is missing a label already defined on the resource in the managed cluster 2) has a label not defined on the resource in the managed cluster.

### **Steps**

1. Create a BareMetalHost resource on `cluster0`:

        $ cat labeltest-baremetalhost.cr.yaml
        ---
        apiVersion: metal3.io/v1alpha1
        kind: BareMetalHost
        metadata:
          name: openshift-master-0
          namespace: openshift-machine-api
          labels:
            "metal3.io/role": "master"
        spec:
          online: true
          bmc:
            address: ipmi://192.168.122.1:6233
            credentialsName: openshift-master-0-secret
        
        $ kubectl --context=cluster0 create -n openshift-machine-api -f labeltest-baremetalhost.cr.yaml

2. Create SyncSet:

        $ cat labeltest-syncset.cr.yaml
        ---
        apiVersion: hive.openshift.io/v1
        kind: SyncSet
        metadata:
          name: labeltest-syncset
          namespace: default
        spec:
          clusterDeploymentRefs:
          - name: cluster0
          resourceApplyMode: Sync
          resources:
            - apiVersion: metal3.io/v1alpha1
              kind: BareMetalHost
              metadata:
                name: openshift-master-0
                namespace: openshift-machine-api
                labels:
                  "metal3.io/cluster": "cluster0"
              spec:
                online: true
                bmc:
                  address: ipmi://192.168.122.1:6233
                  credentialsName: openshift-master-0-secret
        
        $ kubectl --context=hub create -f labeltest-syncset.cr.yaml

### Expected Result

The `openshift-master-0` BareMetalHost in the `openshift-machine-api` namespace to be updated with the new label (without destroying the old).

### Actual Result

    # Before SyncSet
    $ kubectl --context cluster0 get baremetalhosts -n openshift-machine-api -o yaml
    apiVersion: v1
    items:
    - apiVersion: metal3.io/v1alpha1
      kind: BareMetalHost
      metadata:
        creationTimestamp: "2020-02-05T20:37:36Z"
        generation: 1
        labels:
          metal3.io/role: master
        name: openshift-master-0
        namespace: openshift-machine-api
        resourceVersion: "18876"
        selfLink: /apis/metal3.io/v1alpha1/namespaces/openshift-machine-api/baremetalhosts/openshift-master-0
        uid: 69e39ad2-69b1-4d38-bef7-56bd05cda78b
      spec:
        bmc:
          address: ipmi://192.168.122.1:6233
          credentialsName: openshift-master-0-secret
        online: true
    kind: List
    metadata:
      resourceVersion: ""
      selfLink: ""
    
    # Creating the SyncSet
    $ kubectl --context hub create -f labeltest-syncset.cr.yaml
    syncset.hive.openshift.io/labeltest-syncset created
    
    # After SyncSet applies
    $ kubectl --context cluster0 get baremetalhosts -n openshift-machine-api -o yaml
    apiVersion: v1
    items:
    - apiVersion: metal3.io/v1alpha1
      kind: BareMetalHost
      metadata:
        annotations:
          kubectl.kubernetes.io/last-applied-configuration: |
            {"apiVersion":"metal3.io/v1alpha1","kind":"BareMetalHost","metadata":{"annotations":{},"labels":{"metal3.io/cluster":"cluster0"},"name":"openshift-master-0","namespace":"openshift-machine-api"},"spec":{"bmc":{"address":"ipmi://192.168.122.1:6233","credentialsName":"openshift-master-0-secret"},"online":true}}
        creationTimestamp: "2020-02-05T20:37:36Z"
        generation: 1
        labels:
          metal3.io/cluster: cluster0
          metal3.io/role: master
        name: openshift-master-0
        namespace: openshift-machine-api
        resourceVersion: "18899"
        selfLink: /apis/metal3.io/v1alpha1/namespaces/openshift-machine-api/baremetalhosts/openshift-master-0
        uid: 69e39ad2-69b1-4d38-bef7-56bd05cda78b
      spec:
        bmc:
          address: ipmi://192.168.122.1:6233
          credentialsName: openshift-master-0-secret
        online: true
    kind: List
    metadata:
      resourceVersion: ""
      selfLink: ""

## SyncSet Merges Annotations

In this scenario, a BareMetalHost resource already exists on the managed cluster with some annotations. We will create a SyncSet with a BareMetalHost resource that 1) is missing an annotation already defined on the resource in the managed cluster 2) has an annotation not defined on the resource in the managed cluster.

### **Steps**

1. Create a BareMetalHost resource on `cluster0`:

        $ cat annotationtest-baremetalhost.cr.yaml
        ---
        apiVersion: metal3.io/v1alpha1
        kind: BareMetalHost
        metadata:
          name: openshift-master-0
          namespace: openshift-machine-api
          annotations:
            "metal3.io/BareMetalRole": "master"
        spec:
          online: true
          bmc:
            address: ipmi://192.168.122.1:6233
            credentialsName: openshift-master-0-secret
        
        $ kubectl --context=cluster0 create -n openshift-machine-api -f annotationtest-baremetalhost.cr.yaml

2. Create SyncSet:

        $ cat annotationtest-syncset.cr.yaml
        ---
        apiVersion: hive.openshift.io/v1
        kind: SyncSet
        metadata:
          name: annotationtest-syncset
          namespace: default
        spec:
          clusterDeploymentRefs:
          - name: cluster0
          resourceApplyMode: Sync
          resources:
            - apiVersion: metal3.io/v1alpha1
              kind: BareMetalHost
              metadata:
                name: openshift-master-0
                namespace: openshift-machine-api
                annotations:
                  "metal3.io/BareMetalHost": "openshift-machine-api/openshift-master-0"
              spec:
                online: true
                bmc:
                  address: ipmi://192.168.122.1:6233
                  credentialsName: openshift-master-0-secret
        
        $ kubectl --context=hub create -f annotationtest-syncset.cr.yaml

### Expected Result

The `openshift-master-0` BareMetalHost in the `openshift-machine-api` namespace to be updated with the new annotation (without destroying the old).

### Actual Result

    $ kubectl --context cluster0 get baremetalhosts -n openshift-machine-api -o yaml
    apiVersion: v1
    items:
    - apiVersion: metal3.io/v1alpha1
      kind: BareMetalHost
      metadata:
        annotations:
          kubectl.kubernetes.io/last-applied-configuration: |
            {"apiVersion":"metal3.io/v1alpha1","kind":"BareMetalHost","metadata":{"annotations":{"metal3.io/BareMetalHost":"openshift-machine-api/openshift-master-0"},"name":"openshift-master-0","namespace":"openshift-machine-api"},"spec":{"bmc":{"address":"ipmi://192.168.122.1:6233","credentialsName":"openshift-master-0-secret"},"online":true}}
          metal3.io/BareMetalHost: openshift-machine-api/openshift-master-0
          metal3.io/BareMetalRole: master
        creationTimestamp: "2020-02-05T20:49:39Z"
        generation: 1
        name: openshift-master-0
        namespace: openshift-machine-api
        resourceVersion: "19769"
        selfLink: /apis/metal3.io/v1alpha1/namespaces/openshift-machine-api/baremetalhosts/openshift-master-0
        uid: 56623d68-abbe-4b7b-9dfb-89892ad8c662
      spec:
        bmc:
          address: ipmi://192.168.122.1:6233
          credentialsName: openshift-master-0-secret
        online: true
    kind: List
    metadata:
      resourceVersion: ""
      selfLink: ""

## SyncSet Merges Resource.Spec

In this scenario, a BareMetalHost resource already exists on the managed cluster with some annotations. We will create a SyncSet with a BareMetalHost resource that 1) is missing an annotation already defined on the resource in the managed cluster 2) has an annotation not defined on the resource in the managed cluster.

### **Steps**

1. Create a BareMetalHost resource on `cluster0`:

        $ cat spectest-baremetalhost.cr.yaml
        ---
        apiVersion: metal3.io/v1alpha1
        kind: BareMetalHost
        metadata:
          name: openshift-master-0
          namespace: openshift-machine-api
        spec:
          online: true
          bmc:
            address: ipmi://192.168.122.1:6233
            credentialsName: openshift-master-0-secret
          image:
            url: "http://foo.example.io/image/path"
            checksum: "foochecksum"
        
        $ kubectl --context=cluster0 create -n openshift-machine-api -f spectest-baremetalhost.cr.yaml

2. Create SyncSet:

        $ cat spectest-syncset.cr.yaml
        ---
        apiVersion: hive.openshift.io/v1
        kind: SyncSet
        metadata:
          name: spectest-syncset
          namespace: default
        spec:
          clusterDeploymentRefs:
          - name: cluster0
          resourceApplyMode: Sync
          resources:
            - apiVersion: metal3.io/v1alpha1
              kind: BareMetalHost
              metadata:
                name: openshift-master-0
                namespace: openshift-machine-api
              spec:
                online: true
                bmc:
                  address: ipmi://192.168.122.1:6233
                  credentialsName: openshift-master-0-secret
                bootMACAddress: "00:1B:44:11:3A:B7"
                hardwareProfile: "hardwareProfile"
        
        $ kubectl --context=hub create -f spectest-syncset.cr.yaml

### Expected Result

The `openshift-master-0` BareMetalHost in the `openshift-machine-api` namespace to be updated with the new information from the SyncSet (without destroying existing fields).

### Actual Result

    $ kubectl --context cluster0 get baremetalhosts -n openshift-machine-api -o yaml
    apiVersion: v1
    items:
    - apiVersion: metal3.io/v1alpha1
      kind: BareMetalHost
      metadata:
        annotations:
          kubectl.kubernetes.io/last-applied-configuration: |
            {"apiVersion":"metal3.io/v1alpha1","kind":"BareMetalHost","metadata":{"annotations":{},"name":"openshift-master-0","namespace":"openshift-machine-api"},"spec":{"bmc":{"address":"ipmi://192.168.122.1:6233","credentialsName":"openshift-master-0-secret"},"bootMACAddress":"00:1B:44:11:3A:B7","hardwareProfile":"hardwareProfile","online":true}}
        creationTimestamp: "2020-02-05T20:59:33Z"
        generation: 2
        name: openshift-master-0
        namespace: openshift-machine-api
        resourceVersion: "20498"
        selfLink: /apis/metal3.io/v1alpha1/namespaces/openshift-machine-api/baremetalhosts/openshift-master-0
        uid: 1e90326d-a656-4fd7-87f7-7c8c2b98c642
      spec:
        bmc:
          address: ipmi://192.168.122.1:6233
          credentialsName: openshift-master-0-secret
        bootMACAddress: 00:1B:44:11:3A:B7
        hardwareProfile: hardwareProfile
        image:
          checksum: foochecksum
          url: http://foo.example.io/image/path
        online: true
    kind: List
    metadata:
      resourceVersion: ""
      selfLink: ""

**Comment by djzager** [comment]
Every scenario's expected behavior matched the actual behavior. Not sure how this can this be incorporated into documentation but hopefully it is helpful nonetheless.

**Comment by gregsheremeta** [comment]
> Not sure how this can this be incorporated into documentation but hopefully it is helpful nonetheless.

@djzager can you summarize your conclusions in a few sentences, and I'll toss it into the docs?

**Comment by djzager** [comment]
I think a **tl;dr** would look something like:

SyncSets support referencing a Secret in any namespace in `secretMappings` and resources specified will be merged -- just like `kubectl apply` -- with existing resources on the managed cluster.
